internal/k8s/client.go

```diff
@@ -314,42 +339,141 @@ func (k K8sClient) Upsert(ctx context.Context, entities []K8sEntity, timeout tim
+		return nil, nil, err
 	}
 
-	return parseYAMLFromStringWithDeletedResources(stdout)
+	return existing, resources, nil
+}
+
+// Loosely adapted from Helm: visit the resource list and get all the existing resources.
+func existingResourceConflict(resources kube.ResourceList) (kube.ResourceList, error) {
+	var requireUpdate kube.ResourceList
+
+	err := resources.Visit(func(info *resource.Info, err error) error {
+		if err != nil {
+			return err
+		}
+
+		helper := resource.NewHelper(info.Client, info.Mapping)
+		existing, err := helper.Get(info.Namespace, info.Name)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return nil
+			}
+			return errors.Wrap(err, "could not get information about the resource")
+		}
+
+		newRes := *info
+		_ = newRes.Refresh(existing, true)
+
+		requireUpdate.Append(&newRes)
+		return nil
+	})
+
+	return requireUpdate, err
+}
+
+func (k K8sClient) helmResultToEntities(result *kube.Result) ([]K8sEntity, error) {
+	entities := []K8sEntity{}
+	for _, info := range result.Created {
+		entities = append(entities, NewK8sEntity(info.Object))
+	}
+	for _, info := range result.Updated {
+		entities = append(entities, NewK8sEntity(info.Object))
+	}
+
+	// Helm parses the results as unstructured info, but Tilt needs them parsed with the current
+	// API scheme. The easiest way to do this is to serialize them to yaml and re-parse again.
+	buf, err := SerializeSpecYAMLToBuffer(entities)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading k8s result")
+	}
+
+	parsed, err := ParseYAML(buf)
+	if err != nil {
+		return nil, errors.Wrap(err, "parsing k8s result")
+	}
+	return parsed, nil
 }
 
 // applyEntityAndMaybeForce `kubectl apply`'s the given entity, and if the call fails with
 // an immutible field error, attempts to `replace --force` it.
 func (k K8sClient) applyEntityAndMaybeForce(ctx context.Context, entity K8sEntity) ([]K8sEntity, error) {
-	stdout, stderr, err := k.actOnEntity(ctx, []string{"apply", "-o", "yaml"}, entity)
+	existing, resources, err := k.prepareUpdate(ctx, []K8sEntity{entity})
 	if err != nil {
-		reason, shouldTryReplace := maybeShouldTryReplaceReason(stderr)
+		return nil, errors.Wrap(err, "kubectl apply")
+	}
 
+	result, err := k.helmKubeClient.Update(existing, resources, false)
+	if err != nil {
+		reason, shouldTryReplace := maybeShouldTryReplaceReason(err.Error())
 		if !shouldTryReplace {
-			return nil, errors.Wrapf(err, "kubectl apply:\nstderr: %s", stderr)
+			return nil, err
 		}
 
-		// NOTE(maia): we don't use `kubecutl replace --force`, because we want to ensure that all
+		// NOTE(maia): we don't use `kubectl replace --force`, because we want to ensure that all
 		// dependant pods get deleted rather than orphaned. We WANT these pods to be deleted
 		// and recreated so they have all the new labels, etc. of their controlling k8s entity.
 		logger.Get(ctx).Infof("Applying %s failed. Retrying with 'kubectl delete && create': %s", entity.Name(), reason)
```

Comment:
```suggestion
		logger.Get(ctx).Infof("Updating %s failed. Attempting to delete + recreate': %s", entity.Name(), reason)
```

Reply:
or maybe we want to use "kubectl delete", etc, metaphorically?

Reply:
ya, i went back and forth on this -- i think this speaks to the general debate about how much to hide vs expose kubernetes mechanics and when